File: core/time_control.py
"""
CFL 기반 적응 시간 간격 제어.

Courant-Friedrichs-Lewy 조건 및 Fourier 수 기반 자동 dt 조절.
"""

import numpy as np


class AdaptiveTimeControl:
    """CFL 기반 적응 시간 간격 제어기."""

    def __init__(self, dt_init, dt_min=1e-8, dt_max=1.0,
                 cfl_target=0.5, cfl_max=1.0,
                 fourier_max=0.5,
                 growth_factor=1.2, shrink_factor=0.5,
                 safety_factor=0.9):
        self.dt = dt_init
        self.dt_min = dt_min
        self.dt_max = dt_max
        self.cfl_target = cfl_target
        self.cfl_max = cfl_max
        self.fourier_max = fourier_max
        self.growth_factor = growth_factor
        self.shrink_factor = shrink_factor
        self.safety_factor = safety_factor

        # History tracking
        self.dt_history = []
        self.cfl_history = []
        self.fourier_history = []

    def compute_cfl(self, mesh, velocity, dt):
        """
        CFL number per cell: CFL_i = |u_i| * dt / dx_i

        Parameters
        ----------
        mesh : FVMesh
        velocity : np.ndarray (n_cells,) or (n_cells, ndim)
            Cell-center velocities
        dt : float

        Returns
        -------
        cfl_max : float
        cfl_array : np.ndarray (n_cells,)
        """
        ndim = getattr(mesh, 'ndim', 2)
        n = mesh.n_cells

        # Compute characteristic cell size: dx = V^(1/ndim)
        dx = np.array([mesh.cells[i].volume ** (1.0 / ndim) for i in range(n)])
        dx = np.maximum(dx, 1e-30)

        # Velocity magnitude
        velocity = np.asarray(velocity)
        if velocity.ndim == 1:
            u_mag = np.abs(velocity)
        else:
            u_mag = np.linalg.norm(velocity, axis=1)

        cfl_array = u_mag * dt / dx
        return float(np.max(cfl_array)), cfl_array

    def compute_fourier(self, mesh, alpha_diff, dt):
        """
        Fourier number per cell: Fo_i = alpha * dt / dx_i^2

        Parameters
        ----------
        alpha_diff : float or np.ndarray
            Thermal/momentum diffusivity

        Returns
        -------
        fo_max : float
        fo_array : np.ndarray
        """
        ndim = getattr(mesh, 'ndim', 2)
        n = mesh.n_cells
        dx = np.array([mesh.cells[i].volume ** (1.0 / ndim) for i in range(n)])
        dx = np.maximum(dx, 1e-30)

        fo_array = alpha_diff * dt / (dx ** 2)
        return float(np.max(fo_array)), fo_array

    def compute_dt(self, mesh, velocity, alpha_diff=None, converged=True):
        """
        Compute next time step based on CFL and Fourier constraints.

        Returns
        -------
        dt_new : float
        info : dict with 'cfl_max', 'fourier_max', 'dt_limited_by'
        """
        ndim = getattr(mesh, 'ndim', 2)
        n = mesh.n_cells

        dx = np.array([mesh.cells[i].volume ** (1.0 / ndim) for i in range(n)])
        dx = np.maximum(dx, 1e-30)

        # Velocity magnitude
        velocity = np.asarray(velocity)
        if velocity.ndim == 1:
            u_mag = np.abs(velocity)
        else:
            u_mag = np.linalg.norm(velocity, axis=1)

        # CFL constraint: dt_cfl = cfl_target * dx / |u|
        u_safe = np.maximum(u_mag, 1e-30)
        dt_cfl_arr = self.cfl_target * dx / u_safe
        dt_cfl = float(np.min(dt_cfl_arr)) * self.safety_factor

        dt_new = dt_cfl
        limited_by = 'cfl'

        # Fourier constraint
        if alpha_diff is not None:
            dt_fo_arr = self.fourier_max * dx**2 / np.maximum(alpha_diff, 1e-30)
            dt_fo = float(np.min(dt_fo_arr)) * self.safety_factor
            if dt_fo < dt_new:
                dt_new = dt_fo
                limited_by = 'fourier'

        # Growth limit
        dt_new = min(dt_new, self.growth_factor * self.dt)

        # Shrink on divergence
        if not converged:
            dt_new = self.shrink_factor * self.dt
            limited_by = 'divergence'

        # Clip to bounds
        dt_new = float(np.clip(dt_new, self.dt_min, self.dt_max))

        # Compute actual CFL at new dt
        cfl_max, _ = self.compute_cfl(mesh, velocity, dt_new)
        fo_max = 0.0
        if alpha_diff is not None:
            fo_max, _ = self.compute_fourier(mesh, alpha_diff, dt_new)

        # Store
        self.dt = dt_new
        self.dt_history.append(dt_new)
        self.cfl_history.append(cfl_max)
        self.fourier_history.append(fo_max)

        info = {
            'cfl_max': cfl_max,
            'fourier_max': fo_max,
            'dt_limited_by': limited_by,
            'dt': dt_new
        }
        return dt_new, info

    def get_info(self):
        """Return current state."""
        return {
            'dt': self.dt,
            'dt_history': self.dt_history.copy(),
            'cfl_history': self.cfl_history.copy(),
            'fourier_history': self.fourier_history.copy()
        }

File: cpp/src/time_control.cpp
#include "twofluid/time_control.hpp"
#include <algorithm>
#include <cmath>

namespace twofluid {

AdaptiveTimeControl::AdaptiveTimeControl(
    double dt_init, double dt_min, double dt_max,
    double cfl_target, double cfl_max, double fourier_max,
    double growth_factor, double shrink_factor, double safety_factor)
    : dt_(dt_init), dt_min_(dt_min), dt_max_(dt_max),
      cfl_target_(cfl_target), cfl_max_val_(cfl_max),
      fourier_max_(fourier_max),
      growth_factor_(growth_factor), shrink_factor_(shrink_factor),
      safety_factor_(safety_factor) {}

std::pair<double, Eigen::VectorXd>
AdaptiveTimeControl::compute_cfl(const FVMesh& mesh,
                                  const Eigen::VectorXd& u_mag,
                                  double dt) const {
    int n = mesh.n_cells;
    double inv_ndim = 1.0 / mesh.ndim;
    Eigen::VectorXd cfl(n);

    for (int i = 0; i < n; ++i) {
        double dx = std::pow(std::max(mesh.cells[i].volume, 1e-30), inv_ndim);
        cfl(i) = u_mag(i) * dt / dx;
    }
    return {cfl.maxCoeff(), cfl};
}

std::pair<double, Eigen::VectorXd>
AdaptiveTimeControl::compute_fourier(const FVMesh& mesh,
                                      double alpha_diff,
                                      double dt) const {
    int n = mesh.n_cells;
    double inv_ndim = 1.0 / mesh.ndim;
    Eigen::VectorXd fo(n);

    for (int i = 0; i < n; ++i) {
        double dx = std::pow(std::max(mesh.cells[i].volume, 1e-30), inv_ndim);
        fo(i) = alpha_diff * dt / (dx * dx);
    }
    return {fo.maxCoeff(), fo};
}

TimeStepInfo AdaptiveTimeControl::compute_dt(
    const FVMesh& mesh, const Eigen::VectorXd& u_mag,
    double alpha_diff, bool converged) {

    int n = mesh.n_cells;
    double inv_ndim = 1.0 / mesh.ndim;

    // CFL constraint: dt_cfl = cfl_target * dx / |u|
    double dt_cfl = dt_max_;
    for (int i = 0; i < n; ++i) {
        double dx = std::pow(std::max(mesh.cells[i].volume, 1e-30), inv_ndim);
        double u = std::max(u_mag(i), 1e-30);
        double dt_i = cfl_target_ * dx / u;
        dt_cfl = std::min(dt_cfl, dt_i);
    }
    dt_cfl *= safety_factor_;

    double dt_new = dt_cfl;
    std::string limited_by = "cfl";

    // Fourier constraint
    if (alpha_diff > 0.0) {
        double dt_fo = dt_max_;
        for (int i = 0; i < n; ++i) {
            double dx = std::pow(std::max(mesh.cells[i].volume, 1e-30), inv_ndim);
            double dt_i = fourier_max_ * dx * dx / std::max(alpha_diff, 1e-30);
            dt_fo = std::min(dt_fo, dt_i);
        }
        dt_fo *= safety_factor_;
        if (dt_fo < dt_new) {
            dt_new = dt_fo;
            limited_by = "fourier";
        }
    }

    // Growth limit
    dt_new = std::min(dt_new, growth_factor_ * dt_);

    // Shrink on divergence
    if (!converged) {
        dt_new = shrink_factor_ * dt_;
        limited_by = "divergence";
    }

    // Clip
    dt_new = std::max(dt_min_, std::min(dt_new, dt_max_));

    // Compute actual CFL at new dt
    auto [cfl_max, cfl_arr] = compute_cfl(mesh, u_mag, dt_new);
    double fo_max = 0.0;
    if (alpha_diff > 0.0) {
        auto [fm, fa] = compute_fourier(mesh, alpha_diff, dt_new);
        fo_max = fm;
    }

    dt_ = dt_new;
    dt_history.push_back(dt_new);
    cfl_history.push_back(cfl_max);
    fourier_history.push_back(fo_max);

    return {cfl_max, fo_max, limited_by, dt_new};
}

} // namespace twofluid
